router.go

```diff
@@ -36,19 +37,27 @@ type serviceProcedure struct {
 	procedure string
 }
 
+type serviceProcedureEncoding struct {
+	service   string
+	procedure string
+	encoding  transport.Encoding
+}
+
 // MapRouter is a Router that maintains a map of the registered
 // procedures.
 type MapRouter struct {
-	defaultService string
-	entries        map[serviceProcedure]transport.Procedure
+	defaultService            string
+	serviceProcedures         map[serviceProcedure]transport.Procedure
+	serviceProcedureEncodings map[serviceProcedureEncoding]transport.Procedure
```

Comment:
A single `map[serviceProcedure]transport.Procedure` is just analog for `map[service]map[procedure]transport.Procedure`. Two maps make this a lot more complicated than it needs to be. It's cleaner/simpler/more straightforward to have a single map than have two maps and somehow enforce the invariant that a key exists in only one of the two maps. Plus that's two map lookups (in the worst case) at dispatch rather than one.

RE: Interface Indirection: Just an implementation detail. Interface is just one way to do it. Another is a union:

    // With map[serviceProcedure]encodingProcedure,
    type encodingProcedure struct {
        // Only one of the following is set.
        Wildcard transport.Procedure
        PerEncoding map[transport.Encoding]transport.Procedure
    }